### ISSUE TITLE
Service Station Finder Extension disables Dropbox Status Icons

### ISSUE THREAD

**Issue by gordon8214** [knurling/ServiceStation#36]
The Service Station Finder extension seems to cause the Dropbox Finder extension's status indicators on items to disappear.

If both extensions are enabled, Dropbox has no status indicators. If I turn off the Service Station extension, the Dropbox indicators appear. If I then turn Service Station back on, the indicators remain until reboot/logout/Finder restart, at which point they disappear again.

[Video Demo](https://www.dropbox.com/s/ar9mult9f26xqmh/Service%20Station%20Dropbox%20Issue.mp4?raw=1)

**Comment by VKondakoff** [comment]
I can confirm the issue. SS Version 2020.5 (18), Catalina 10.15.4.

**Comment by pkamb** [comment]
This will unfortunately most likely be `cantfix` for the current versions of macOS.

Service Station doesn't do *any* badging work as a Finder Extension, which corresponds with advice given by Apple:

 > For example, I would avoid doing ANY icon badging, since it’s very likely to interfere with other FinderSync apps.

I also don't have a good way to exclude the Dropbox folder from the list of Service Station-enabled folders (which you might not want to do anyway).

It appears that macOS is simply picking a single Finder Extension to be responsible for badges in any given folder. If you enable Dropbox and then Service Station, Dropbox wins. But if Service Station is enabled first, Dropbox can't add its icons. I'm not sure what heuristic is being used, but it's unfortunately not alphabetical :(

This will probably need an OS update from Apple. I will log a bug and post it here, and any duplicate bugs you file will help.

It would probably also work to somehow script your system to periodically disable both Finder Extensions and re-enable them in the correct order. I won't be able to do that from within the sandboxed Service Station app, but it could be done from a un-sandboxed context

**Comment by gordon8214** [comment]
Did you ever file a bug for this? Do you have the number?

**Comment by pkamb** [comment]
> ### Finder Sync Extension badges not shown if multiple Finder Sync Extensions are enabled.
 > 
 > DUPLICATE OF FB7711264
 > 
 > I'm using the Finder Sync Extension app Service Station: https://apps.apple.com/us/app/service-station/id1503136033?ls=1/
 > 
 > I also have Dropbox installed, and the Dropbox Finder Extension enabled.
 > 
 > Service Station does not show any badge icons on Finder items. Dropbox does.
 > 
 > But if you enable Service Station BEFORE Dropbox, Dropbox will not be able to show any Finder badge icons.
 > 
 > ### Reproduction steps:
 > 
 > 1. System Preferences > Extensions > Finder Extensions.
 > 2. Enable Service Station Finder Extension, then Dropbox Finder Extension
 > 3. Observe your dropbox folder at `~/Dropbox/`
 > 
 > ### Result:
 > 
 >  - Dropbox syncing icons are NOT shown on Finder icons in Dropbox.
 > 
 > ### Expected Result.
 > 
 >  - Dropbox folder icons should be shown.
 >  - Service Station does not use Finder icons. Dropbox should thus be allowed to.
 > 
 > ### Fix:
 > 
 > If you reverse the order and Enable Dropbox PRIOR TO enabling Service Station, the Dropbox finder icons are shown.
 > 
 > 1. Disable Service Station extension.
 > 2. Note that Dropbox icons are now shown!
 > 3. Enable Service Station extension.
 > 4. Dropbox icons are still shown.
 > 
 > ### Notes:
 > 
 > Service Station's `FIFinderSyncController.default().directoryURLs` directory contains a parent directory of the Dropbox folder.
 > 
 > Apparently order matters. Only the first Finder Sync Extension to be enabled will be able to display badges, even if that extension does not ever show badges.
 > 
 > This should be changed to instead show badges for the first Extension that attempts to Display Finder icon badge icons.

**Comment by gingerbeardman** [comment]
this feels like a namespace clash, could it be?

**Comment by pkamb** [comment]
What namespace? The apps have different BundleIDs...

**Comment by pkamb** [comment]
I've opened an [Apple DTS ticket](https://developer.apple.com/support/technical/) for this issue; hopefully an Apple engineer can help with an explanation and/or workaround.

**Comment by gingerbeardman** [comment]
Yeah, I have no idea. :)

**Comment by moffat** [comment]
In case anybody finds this useful, I solved this problem using a Keyboard Maestro Macro that is triggered every time Finder launches. It waits a few seconds (to make sure finder is running) and then deactivates and activates the ServiceStation finder extension with:

```
pluginkit -e ignore -i com.knurling.ServiceStation.FinderSync
pluginkit -e use -i com.knurling.ServiceStation.FinderSync
```

The macro is attached. 

[Service Station Fix.kmmacros.zip](https://github.com/knurling/ServiceStation/files/5013276/Service.Station.Fix.kmmacros.zip)

**Comment by gordon8214** [comment]
> In case anybody finds this useful, I solved this problem (I'm running using a Keyboard Maestro Macro that is triggered every time Finder launches. It waits to make sure finder is running and then deactivates and activates the ServiceStation finder extension with:
> 
> ```
> pluginkit -e ignore -i com.knurling.ServiceStation.FinderSync
> pluginkit -e use -i com.knurling.ServiceStation.FinderSync
> ```
> 
> The macro is attached.
> 
> [Service Station Fix.kmmacros.zip](https://github.com/knurling/ServiceStation/files/5013306/Service.Station.Fix.kmmacros.zip)

Great workaround. I found I had to add `sleep 1` between the commands in order for it to work reliably though.

**Comment by luckman212** [comment]
I don't think this is a problem anymore with the new FileProvider based extensions for OneDrive, Dropbox, Google Drive etc? Are you guys still seeing this issue on recent versions of macOS?

**Comment by EddyRezumi** [comment]
> I don't think this is a problem anymore with the new FileProvider based extensions for OneDrive, Dropbox, Google Drive etc? Are you guys still seeing this issue on recent versions of macOS?

I have OneDrive on macOS Monterey and badges are showing up just fine. But instead, I don't have the app I set up in the contextual menu when I right click a file. The Finder doesn't allow 2 extensions to change it's right click menu ?

EDIT : I tried to deactivate OneDrive extension, did not help. And actually we can't have it disabled, since it is necessary for synchronization.
